daffodil-core/src/main/scala/org/apache/daffodil/grammar/GrammarTerm.scala

```diff
@@ -94,16 +94,19 @@ abstract class Gram(contextArg: SchemaComponent)
    * Note: This should not evaluate the argument unless it has to.
    */
   def ~(qq: => Gram) = {
-    lazy val q = qq.deref
+    val q = qq.deref
     val self = this.deref
 
     val res =
       if (self.isEmpty) {
-        if (q.isEmpty)
+        if (q.isEmpty) {
           EmptyGram
-        else q
-      } else if (q.isEmpty) self
-      else {
+        } else {
+          q
+        }
+      } else if (q.isEmpty) {
+        self
```

Comment:
I think this is the only change here? I'm not sure I understandt this change. What issue does this fix?